# <font color='blue'>Data Science Academy</font>

## <font color='blue'>Fundamentos de Linguagem Python Para Análise de Dados e Data Science</font>

## <font color='blue'>Manipulação de Dados com Pandas</font>

![DSA](imagens/dsa_cap10.png)

In [ ]:
# Versão da Linguagem Python
from platform import python_version
print('Versão da Linguagem Python Usada Neste Jupyter Notebook:', python_version())

Guia de Usuário do Pandas:

https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html#user-guide

In [ ]:
# Instala a versão exata do pacote
%pip install pandas

In [1]:
import pandas as pd

In [2]:
pd.__version__

'2.2.2'

## Manipulando Dados em DataFrames do Pandas 

In [4]:
# Cria um dicionário
dados = {'Estado': ['Santa Catarina', 'Rio de Janeiro', 'Tocantins', 'Bahia', 'Minas Gerais'], 
         'Ano': [2004, 2005, 2006, 2007, 2008], 
         'Taxa Desemprego': [1.5, 1.7, 1.6, 2.4, 2.7]}

In [5]:
print(dados)

{'Estado': ['Santa Catarina', 'Rio de Janeiro', 'Tocantins', 'Bahia', 'Minas Gerais'], 'Ano': [2004, 2005, 2006, 2007, 2008], 'Taxa Desemprego': [1.5, 1.7, 1.6, 2.4, 2.7]}


In [6]:
# Importa a função DataFrame do Pandas
from pandas import DataFrame

In [7]:
# Converte o dicionário em um dataframe
df = DataFrame(dados)

In [8]:
# Visualiza as 5 primeiras linhas
df.head()

Estado   Ano  Taxa Desemprego
0  Santa Catarina  2004              1.5
1  Rio de Janeiro  2005              1.7
2       Tocantins  2006              1.6
3           Bahia  2007              2.4
4    Minas Gerais  2008              2.7

In [10]:
type(df)

pandas.core.frame.DataFrame

In [11]:
# Reorganizando as colunas
DataFrame(dados, columns = ['Estado', 'Taxa Desemprego', 'Ano'])

Estado  Taxa Desemprego   Ano
0  Santa Catarina              1.5  2004
1  Rio de Janeiro              1.7  2005
2       Tocantins              1.6  2006
3           Bahia              2.4  2007
4    Minas Gerais              2.7  2008

In [12]:
# Criando outro dataframe com os mesmo dados anteriores mas adicionando uma coluna
df2 = DataFrame(dados, 
                columns = ['Estado', 'Taxa Desemprego', 'Taxa Crescimento', 'Ano'], 
                index = ['estado1', 'estado2', 'estado3', 'estado4', 'estado5'])

In [13]:
df2

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado1  Santa Catarina              1.5              NaN  2004
estado2  Rio de Janeiro              1.7              NaN  2005
estado3       Tocantins              1.6              NaN  2006
estado4           Bahia              2.4              NaN  2007
estado5    Minas Gerais              2.7              NaN  2008

In [17]:
df2.values

array([['Santa Catarina', 1.5, nan, 2004],
       ['Rio de Janeiro', 1.7, nan, 2005],
       ['Tocantins', 1.6, nan, 2006],
       ['Bahia', 2.4, nan, 2007],
       ['Minas Gerais', 2.7, nan, 2008]], dtype=object)

In [18]:
df2.dtypes

Estado               object
Taxa Desemprego     float64
Taxa Crescimento     object
Ano                   int64
dtype: object

In [20]:
df2.columns

Index(['Estado', 'Taxa Desemprego', 'Taxa Crescimento', 'Ano'], dtype='object')

In [21]:
# Imprimindo apenas uma coluna do Dataframe
df2['Estado']

estado1    Santa Catarina
estado2    Rio de Janeiro
estado3         Tocantins
estado4             Bahia
estado5      Minas Gerais
Name: Estado, dtype: object

In [ ]:
# Linguagem Python é case sensitive
# df2['estado']

In [22]:
# Imprimindo apenas duas colunas do Dataframe
df2[ ['Taxa Desemprego', 'Ano'] ]

Taxa Desemprego   Ano
estado1              1.5  2004
estado2              1.7  2005
estado3              1.6  2006
estado4              2.4  2007
estado5              2.7  2008

In [23]:
df2.index

Index(['estado1', 'estado2', 'estado3', 'estado4', 'estado5'], dtype='object')

In [24]:
# Filtrando pelo índice
df2.filter(items = ['estado3'], axis = 0)

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado3  Tocantins              1.6              NaN  2006

## Usando NumPy e Pandas Para Manipulação de Dados

In [25]:
df2.head()

Estado  Taxa Desemprego Taxa Crescimento   Ano
estado1  Santa Catarina              1.5              NaN  2004
estado2  Rio de Janeiro              1.7              NaN  2005
estado3       Tocantins              1.6              NaN  2006
estado4           Bahia              2.4              NaN  2007
estado5    Minas Gerais              2.7              NaN  2008

In [26]:
df2.dtypes

Estado               object
Taxa Desemprego     float64
Taxa Crescimento     object
Ano                   int64
dtype: object

In [27]:
# Resumo estatístico do Dataframe
df2.describe()

Taxa Desemprego          Ano
count         5.000000     5.000000
mean          1.980000  2006.000000
std           0.535724     1.581139
min           1.500000  2004.000000
25%           1.600000  2005.000000
50%           1.700000  2006.000000
75%           2.400000  2007.000000
max           2.700000  2008.000000

In [31]:
df2.isna()

Estado  Taxa Desemprego  Taxa Crescimento    Ano
estado1   False            False              True  False
estado2   False            False              True  False
estado3   False            False              True  False
estado4   False            False              True  False
estado5   False            False              True  False

In [30]:
df2['Taxa Crescimento'].isna()

estado1    True
estado2    True
estado3    True
estado4    True
estado5    True
Name: Taxa Crescimento, dtype: bool

In [32]:
# Importando o NumPy
import numpy as np

In [35]:
# Usando o NumPy para alimentar uma das colunas do dataframe
df2['Taxa Crescimento'] = np.arange(5.)

In [36]:
df2

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007
estado5    Minas Gerais              2.7               4.0  2008

In [37]:
df2.dtypes

Estado               object
Taxa Desemprego     float64
Taxa Crescimento    float64
Ano                   int64
dtype: object

In [38]:
df2['Taxa Crescimento'].isna()

estado1    False
estado2    False
estado3    False
estado4    False
estado5    False
Name: Taxa Crescimento, dtype: bool

In [39]:
# Resumo estatístico do Dataframe
df2.describe()

Taxa Desemprego  Taxa Crescimento          Ano
count         5.000000          5.000000     5.000000
mean          1.980000          2.000000  2006.000000
std           0.535724          1.581139     1.581139
min           1.500000          0.000000  2004.000000
25%           1.600000          1.000000  2005.000000
50%           1.700000          2.000000  2006.000000
75%           2.400000          3.000000  2007.000000
max           2.700000          4.000000  2008.000000

## Slicing de DataFrames do Pandas

In [40]:
df2

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007
estado5    Minas Gerais              2.7               4.0  2008

In [41]:
df2['estado2':'estado4']

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006
estado4           Bahia              2.4               3.0  2007

In [42]:
df2[ df2['Taxa Desemprego'] < 2 ]

Estado  Taxa Desemprego  Taxa Crescimento   Ano
estado1  Santa Catarina              1.5               0.0  2004
estado2  Rio de Janeiro              1.7               1.0  2005
estado3       Tocantins              1.6               2.0  2006

In [43]:
df2['Taxa Crescimento']

estado1    0.0
estado2    1.0
estado3    2.0
estado4    3.0
estado5    4.0
Name: Taxa Crescimento, dtype: float64

In [44]:
df2[['Estado', 'Taxa Crescimento']]

Estado  Taxa Crescimento
estado1  Santa Catarina               0.0
estado2  Rio de Janeiro               1.0
estado3       Tocantins               2.0
estado4           Bahia               3.0
estado5    Minas Gerais               4.0

In [ ]:
df2[['Estado', 'Taxa Crescimento', 'Ano']]

## Preenchendo Valores Ausentes em DataFrames do Pandas

A função fillna() é usada para preencher os valores ausentes. A função oferece muitas opções. Podemos usar um valor específico, uma função agregada (por exemplo, média) ou o valor anterior ou seguinte.

Para esse exemplo usaremos a moda, a estatística que representa o valor que aparece mais vezes em uma variável.

In [45]:
# Primeiro importamos um dataset
dsa_df = pd.read_csv("dataset.csv")

In [46]:
dsa_df.head(15)

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
0   CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
1   CA-2016-152156  2016-11-08   CG-12520     Consumer  United States   
2   CA-2016-138688  2016-06-12   DV-13045    Corporate  United States   
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
4   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
6   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
7   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
8   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
11  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
12  CA-2017-114412  2017-04-15   AA-10480     Consumer  United States   
13  CA-2016-161389  2016-12-05   IM-15070     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   

     Regiao       ID_Produto        Categoria  \
0     South  FUR-BO-10001798        Furniture   
1     South  FUR-CH-10000454        Furniture   
2      West  OFF-LA-10000240  Office Supplies   
3     South  FUR-TA-10000577        Furniture   
4     South  OFF-ST-10000760  Office Supplies   
5      West  FUR-FU-10001487        Furniture   
6      West  OFF-AR-10002833  Office Supplies   
7      West  TEC-PH-10002275       Technology   
8      West  OFF-BI-10003910  Office Supplies   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
11     West  TEC-PH-10002033       Technology   
12    South  OFF-PA-10002365  Office Supplies   
13     West  OFF-BI-10003656  Office Supplies   
14  Central  OFF-AP-10002311  Office Supplies   

                                         Nome_Produto  Valor_Venda  Quantidade  
0                   Bush Somerset Collection Bookcase     261.9600         NaN  
1   Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2   Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                      Eldon Fold 'N Roll Cart System      22.3680         2.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
6                                          Newell 322       7.2800         4.0  
7                      Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
8   DXL Angle-View Binders with Locking Rings by S...      18.5040         3.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11      Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
12                                         Xerox 1967      15.5520         3.0  
13        Fellowes PB200 Plastic Comb Binding Machine     407.9760         3.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0

In [50]:
dsa_df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [53]:
# Extraímos a moda da coluna Quantity
moda = dsa_df['Quantidade'].value_counts().index[0]

A moda em Estatística é uma medida de tendência central que representa o valor mais frequente em um conjunto de dados. 

A moda é especialmente útil quando queremos saber qual é o valor mais comum ou popular em um conjunto de dados, seja em uma distribuição unimodal (com apenas uma moda) ou em uma distribuição bimodal (com duas modas).

No entanto, a moda pode não ser tão representativa quanto outras medidas de tendência central, como a média e a mediana, especialmente em distribuições assimétricas ou quando há valores extremos. Por essa razão, é importante analisar diferentes medidas de tendência central e usar a que melhor se adequa aos objetivos da análise estatística.

In [54]:
print(moda)

3.0


In [55]:
# E por fim preenchemos os valores NA com a moda
dsa_df['Quantidade'].fillna(value = moda, inplace = True)

/tmp/ipykernel_10757/4239898850.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dsa_df['Quantidade'].fillna(value = moda, inplace = True)


In [56]:
dsa_df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

## Query (Consulta) de Dados no DataFrame do Pandas

Com o Pandas criamos dataframes, que são essencialmente tabelas. Como tal, podemos fazer consultas, ou simplesmente queries. E para isso usamos o método query(). Veja o exemplo abaixo:

In [57]:
dsa_df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [58]:
# Checamos os valores mínimo e máximo da coluna Valor_Venda
dsa_df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

O intervalo de vendas (Valor_Venda) é de 0.44 a 22638. Vamos fazer uma consulta e retornar todas as vendas dentro de um range de valores. Fazemos isso com a instrução abaixo:

In [59]:
# Geramos um novo dataframe apenas com o intervalo de vendas entre 229 e 10000
df2 = dsa_df.query('229 < Valor_Venda < 10000')

In [60]:
# Então confirmamos os valores mínimo e máximo
df2.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [61]:
# Geramos um novo dataframe apenas com os valores de venda acima da média
df3 = df2.query('Valor_Venda > 766')

In [62]:
df3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

Consulta executada com sucesso!

## Verificando a Ocorrência de Diversos Valores em Uma Coluna

Em nosso conjunto de dados de exemplo temos a coluna Quantidade que representa a quantidade de itens vendidos em cada uma das vendas. Imagine que precisamos saber em quais vendas foram vendidos 5, 7, 9 ou 11 itens. 

Como aplicaríamos esse tipo de filtro ao nosso dataframe?

Fácil. O Pandas oferece o método isin() para checar diversos valores em uma coluna. Quem conhece Linguagem SQL já deve ter percebido que o método é o mesmo que a cláusula IN em SQL. Vamos ao exemplo.

In [63]:
dsa_df.shape

(9994, 11)

In [66]:
# Então aplicamos o filtro
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

Na instrução acima estamos filtrando o dataframe chamado df, retornando todas as linhas onde a coluna Quantidade for igual aos valores 5, 7, 9 ou 11. Passamos uma lista de valores como argumento para o método isin().

Vamos deixar um pouquinho mais divertido. Se você executou a instrução acima, percebeu que foram retornadas 2.128 linhas. E se quisermos retornar somente 10 linhas? É só fatiar o resultado assim:

In [70]:
# Shape
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])].shape

(2128, 11)

In [74]:
dsa_df[dsa_df['Quantidade'].isin([5,7,9,11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0

O Pandas não é incrível? 😁

## Operadores Lógicos Para Manipulação de Dados com Pandas

Os operadores lógicos são excelentes para filtrar dataframes e retornar exatamente os dados que precisamos para nosso trabalho. Para conhecer mais sobre as regras dos operadores lógicos, acesse aqui:

https://pt.wikipedia.org/wiki/Operador_l%C3%B3gico

Primeiro usaremos o operador lógico AND para checar duas condições. Serão retornados os registros quando as duas condições forem verdadeiras.

In [ ]:
# Filtrando as vendas que ocorreram para o segmento de Home Office e na região South
dsa_df[ (dsa_df.Segmento == 'Home Office') & (dsa_df.Regiao == 'South') ].head()

Mas pode ser necessário checar duas condições e retornar os registros se pelo menos uma for verdadeira. Nesse caso usamos o operador OR, conforme abaixo.

In [ ]:
# Filtrando as vendas que ocorreram para o segmento de Home Office ou região South
dsa_df[(dsa_df.Segmento == 'Home Office') | (dsa_df.Regiao == 'South')].tail()

O operador de negação é o contrário do primeiro exemplo.

In [ ]:
# Filtrando as vendas que não ocorreram para o segmento de Home Office e nem na região South
dsa_df[(dsa_df.Segmento != 'Home Office') & (dsa_df.Regiao != 'South')].sample(5)

## Agrupamento de Dados em DataFrames com Group By

A função Pandas Groupby é uma função versátil e fácil de usar que ajuda a obter uma visão geral dos dados. Isso torna mais fácil explorar o conjunto de dados e revelar os relacionamentos entre as variáveis.

O código a seguir agrupará as linhas com base nas combinações Segmento/Regiao/Valor_Venda e nos dará a taxa média de vendas de cada grupo.

In [ ]:
# Aplicamos o group by
dsa_df[['Segmento', 'Regiao', 'Valor_Venda']].groupby(['Segmento', 'Regiao']).mean()

Na instrução acima, primeiro filtramos os dados extraindo 3 colunas: ['Segmento','Regiao','Valor_Venda']. Na sequência, agrupamos por duas colunas: ['Segmento','Regiao']. E então calculamos a média para a coluna que ficou foram do group by, nesse caso a coluna Sales.

O comportamento do group by com Pandas é o mesmo observado na Linguagem SQL.

## Agregação Múltipla com Group By

Vamos explorar mais a função groupby() pois temos diversas opções de sumarização dos dados de forma simples. No exemplo abaixo uniremos a função groupby() com a função agg() para realiza agregação múltipla.

In [ ]:
# Aplicamos o group by
dsa_df[['Segmento', 'Regiao', 'Valor_Venda']].groupby(['Segmento', 'Regiao']).agg(['mean', 'std', 'count'])

Na instrução acima, primeiro filtramos os dados extraindo 3 colunas: ['Segmento','Regiao','Valor_Venda']. Na sequência, agrupamos por duas colunas: ['Segmento','Regiao']. E então agregamos os dados calculando a média, desvio padrão e contagem de elementos para a coluna que ficou fora do group by, nesse caso a coluna Valor_Venda.

A função agg() recebe como argumento uma lista de funções para agregação.

## Filtrando DataFrame do Pandas com Base em Strings

O Pandas oferece diversas funções para manipulação de strings. Começaremos com o filtros de strings com base nas letras iniciais e finais.

In [ ]:
dsa_df.head()

In [ ]:
# Filtramos o dataframe pela coluna Segmento com valores que iniciam com as letras 'Con'
dsa_df[dsa_df.Segmento.str.startswith('Con')].head()

In [ ]:
dsa_df.Segmento.value_counts()

In [ ]:
# Filtramos o dataframe pela coluna Segmento com valores que terminam com as letras 'mer'
dsa_df[dsa_df.Segmento.str.endswith('mer')].head()

As funções startswith() e endswith() são muito úteis quando for necessário filtrar strings por caracteres que apareçam no começo e/ou final.

## Split de Strings em DataFrames do Pandas

Com Pandas podemos realizar diversas tarefas de split de strings dividindo uma coluna ou extraindo elementos do nosso interesse. Vamos ao exemplo!

In [ ]:
dsa_df.head()

In [ ]:
dsa_df['ID_Pedido'].head()

Este é o formato dos dados da coluna "ID_Pedido":

- CA-2016-152156
- US-2015-108966

Temos o país, o ano e o id do pedido. Vamos dividir essa coluna e extrair o ano para gravar em uma nova coluna:

In [ ]:
# Split da coluna pelo caracter '-'
dsa_df['ID_Pedido'].str.split('-')

Observe que o resultado são as listas em Python. Para extrair o ano precisamos especificar o índice da posição que queremos extrair (em nosso caso a posição 2, logo, índice 1 em Python):

In [ ]:
dsa_df['ID_Pedido'].str.split('-').str[1].head()

In [ ]:
# Fazemos o split da coluna e extraímos o item na posição 2 (índice 1)
dsa_df['Ano'] = dsa_df['ID_Pedido'].str.split('-').str[1]

In [ ]:
# Então conferimos a nova coluna criada
dsa_df.head()

## Strip de Strings em DataFrames do Pandas

Cuidado para não confundir. Vimos o Split e agora veremos o Strip. São funções diferentes.

O Split divide a string. O Strip remove caracteres da string. Veja os exemplos.

In [ ]:
dsa_df.head(3)

In [ ]:
dsa_df['Data_Pedido'].head(3)

A coluna 'Data_Pedido' é a data de envio do produto no formato YYYY-MM-DD. Imagine que seja necessário deixar o ano apenas com 2 dígitos sem alterar o tipo da variável. Fazemos isso com a função lstrip(), ou seja, left strip.

In [ ]:
# Vamos remover os dígitos 2 e 0 à esquerda do valor da variável 'Data_Pedido'
dsa_df['Data_Pedido'].str.lstrip('20')

Como não usamos o inplace = True a mudança é somente na memória e não altera o dataframe. Podemos usar ainda as funções rstrip() e strip() com diferentes variações de strip de strings.

In [ ]:
dsa_df['Data_Pedido'].head(3)

## Replace de Strings em DataFrames do Pandas

Se for necessário substituir caracteres dentro de uma string o Pandas oferece uma função para isso também.

Por exemplo, vamos substituir 2 caracteres em uma das colunas.

In [ ]:
dsa_df.head()

In [ ]:
# Substituímos os caracteres CG por AX na coluna 'ID_Cliente'
dsa_df['ID_Cliente'] = dsa_df['ID_Cliente'].str.replace('CG', 'AX')

In [ ]:
dsa_df.head()

E pronto. Fácil assim.

## Combinação de Strings em DataFrames do Pandas

A função cat() pode ser usada para concatenar strings em um dataframe do Pandas.

Vamos criar uma nova coluna concatenando as colunas “ID_Pedido” e “Segmento” com o separador “-”.

In [ ]:
dsa_df.head()

In [ ]:
# Concatenando strings
dsa_df['Pedido_Segmento'] = dsa_df['ID_Pedido'].str.cat(dsa_df['Segmento'], sep = '-')

In [ ]:
dsa_df.head()

## Construção de Gráficos a Partir de DataFrames do Pandas

Vimos até aqui diversas funcionalidades do Pandas que tornam o processo de manipulação de dados realmente simples. E para concluir este capítulo vamos estudar as opções que o Pandas oferece para criação de gráficos diretamente a partir de dataframes, sem a necessidade de usar qualquer outra biblioteca. 

Acompanhe os exemplos.

In [ ]:
# Instala a versão exata do Scikit-learn
!pip install -q scikit-learn==1.2.1

In [ ]:
import sklearn
sklearn.__version__

In [ ]:
# Vamos começar importando o dataset iris do Scikit-learn
from sklearn.datasets import load_iris
data = load_iris()

In [ ]:
# E então carregamos o dataset iris como dataframe do Pandas
import pandas as pd
df = pd.DataFrame(data['data'], columns = data['feature_names'])
df['species'] = data['target']
df.head()

In [ ]:
# Para criar um gráfico de linhas com todas as variáveis do dataframe, basta fazer isso:
df.plot()

In [ ]:
# Que tal um scatter plot com duas variáveis? 
df.plot.scatter(x = 'sepal length (cm)', y = 'sepal width (cm)')

In [ ]:
# E mesmo gráficos mais complexos, como um gráfico de área, pode ser criado:
columns = ['sepal length (cm)', 'petal length (cm)', 'petal width (cm)', 'sepal width (cm)']
df[columns].plot.area()

In [ ]:
# Calculamos a média das colunas agrupando pela coluna species e criamos um gráfico de barras com o resultado
df.groupby('species').mean().plot.bar()

In [ ]:
# Ou então, fazemos a contagem de classes da coluna species e plotamos em um gráfico de pizza
df.groupby('species').count().plot.pie(y = 'sepal length (cm)')

A lista de possibilidades é imensa. Aqui tem muitos outros exemplos:

https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


In [ ]:
# Gráfico KDE (Kernel Density Function) para cada variável do dataframe
df.plot.kde(subplots = True, figsize = (8,8))

In [ ]:
# Boxplot de cada variável numérica
columns = ['sepal length (cm)', 'petal length (cm)', 'petal width (cm)', 'sepal width (cm)']
df[columns].plot.box(figsize = (8,8))

O Pandas é uma verdadeira caixa de ferramentas para manipulação de dados em Python.

# Fim

### Obrigado

### Visite o Blog da Data Science Academy - <a href="http://blog.dsacademy.com.br">Blog DSA</a>